src/hypothesis/core.py

```diff
@@ -153,22 +156,25 @@ def wrapped_test(*arguments, **kwargs):
         return invalid(
             'given must be called with at least one argument')
 
-    if (generator_arguments and (original_argspec.varargs or
-                                 original_argspec.varkw)):
+    if generator_arguments and any([original_argspec.varargs,
+                                    original_argspec.varkw,
+                                    original_argspec.kwonlyargs]):
         return invalid(
-            'varargs or keywords are not supported with positional '
-            'arguments to @given'
+            'positional arguments to @given are not supported with varargs, '
+            'varkeywords, or keyword-only arguments'
         )
 
-    if (
-        len(generator_arguments) > len(original_argspec.args)
-    ):
+    if len(generator_arguments) > len(original_argspec.args):
         return invalid((
             'Too many positional arguments for %s() (got %d but'
             ' expected at most %d') % (
                 name, len(generator_arguments),
                 len(original_argspec.args)))
 
+    if infer in generator_arguments:
```

Comment:
I feel like the "never gets applied to the wrong thing by accident" is about equally likely to happen as with positional arguments to given.

I'm perfectly happy to leave it out for implementation complexity reasons (you've done quite enough work here as it is without me adding yet more!). I anticipate people complaining, but I'm happy to wait for them to complain before thinking further about whether we can be bothered to do the work. :-)